web3/_utils/async_tools.py

```diff
@@ -0,0 +1,95 @@
+import asyncio
+from concurrent.futures import (
+    ThreadPoolExecutor,
+)
+from functools import (
+    singledispatch,
+    wraps,
+)
+import inspect
+import types
+from typing import (
+    Any,
+    Callable,
+    Generator,
+)
+
+_DEFAULT_POOL = ThreadPoolExecutor()
+
+
+def threadpool(f, executor=None):
+    @wraps(f)
+    def wrap(*args, **kwargs):
+        return (executor or _DEFAULT_POOL).submit(f, *args, **kwargs)
+    return wrap
+
+
+@singledispatch
+@threadpool
+def run_in_loop_in_thread(co: Any, args, kwargs):
+    raise TypeError('Called with unsupported argument: {}'.format(co))
+
+
+@run_in_loop_in_thread.register(asyncio.Future)
+@run_in_loop_in_thread.register(types.GeneratorType)
+@run_in_loop_in_thread.register(types.CoroutineType)
+@threadpool
+def run_co_in_new_loop_in_thread(co: Generator[Any, None, Any]) -> Any:
+    try:
+        #  in get_event_loog, self._local._set_called == True, even
+        #  though there is no event_loop.
+        loop = asyncio.get_event_loop()
+    except RuntimeError:
+        loop = asyncio.new_event_loop()
+        #  I get an error regarding what seems like a stale loop.
```

Comment:
I noticed this error when another thread is running something in the background. eg `loop.run_forever`. When another thread tries to run something using the same event loop it gives a `RuntimeError`.